### DISCUSSION TITLE
Akka.Persistence.MongoDb version update causes NULLRef issues

### DISCUSSION THREAD

**Discussion by merthamurcu** [akkadotnet/akka.net#5808] [Q&A]
I am trying to update from netstandard2.0 to net6.0, we were using akka 1.4.5 and trying to update to 1.4.31

some actors throwing nullref due to missing "Ordering" field in the "EventJournal" entries.
we have too many records in the mongoDB to add ordering to each of them realistically, 
is there any work around to fix this error and upgrade my projects safely?

I believe it is introduced in this commit
https://github.com/akkadotnet/Akka.Persistence.MongoDB/commit/bb0f4c6d61931479c2c0dd6ff18296b92f5bd1fb
and line 392 where it throws the nullref

thanks in advance!

**Comment by Aaronontheweb** [comment]
Thanks for letting us know @merthamurcu - we'll take a look at this.

**Comment by Arkatufus** [comment]
The `Ordering` and `Tags` field is added in this PR: https://github.com/akkadotnet/Akka.Persistence.MongoDB/commit/e72b425663d971a211a921b6fe2567e150ac031a

I think there's no way to upgrade without converting the old data structure to the new one

**Reply by Aaronontheweb** [reply]
We can't fix this issue without additional data - post the _exact_ stack trace and a copy of one of your failed data structures please. We introduced the `Ordering` field several years ago.

cc @merthamurcu

**Reply by merthamurcu** [reply]
I believe it was not mandatory before, now it is a breaking change?

2022-04-07 14:29:36.333 +01:00 [Error] Persistence failure when replaying events for persistenceId [FooController]. Last known sequence number [0]
System.NullReferenceException: Object reference not set to an instance of an object.
   at Akka.Persistence.MongoDb.Journal.MongoDbJournal.ToTicks(BsonTimestamp bson)
   at Akka.Persistence.MongoDb.Journal.MongoDbJournal.ToPersistenceRepresentation(JournalEntry entry, IActorRef sender)
   at Akka.Persistence.MongoDb.Journal.MongoDbJournal.<>c__DisplayClass10_0.<ReplayMessagesAsync>b__4(JournalEntry doc)
   at System.Collections.Generic.List`1.ForEach(Action`1 action)
   at Akka.Persistence.MongoDb.Journal.MongoDbJournal.ReplayMessagesAsync(IActorContext context, String persistenceId, Int64 fromSequenceNr, Int64 toSequenceNr, Int64 max, Action`1 recoveryCallback)


{
    "_id" : "FooControllerActor_1",
    "PersistenceId" : "FooControllerActor",
    "SequenceNr" : NumberLong(1),
    "IsDeleted" : false,
    "Payload" : {
[removed]
    },
    "Manifest" : ""
}

**Reply by Aaronontheweb** [reply]
@Arkatufus per what I said in Slack - I think just `??` the `Ordering` with a default value when none is deserialized is a harmless way to handle this case.

**Reply by Arkatufus** [reply]
@merthamurcu We've implemented a small patch here: https://www.nuget.org/packages/Akka.Persistence.MongoDb/1.4.38-beta1
I hope this helps.

**Reply by Aaronontheweb** [reply]
@merthamurcu have you had a chance to try that beta?

**Reply by merthamurcu** [reply]
sorry about that, I didn't have a chance to check. in the middle of dealing with other errors with upgrade :( missing fields on old payload causing us so much trouble, trying to give them null like defaults with conventions. not as straight forward as it sounds

**Reply by Aaronontheweb** [reply]
What other fields are you missing on the payload? And have you seen https://github.com/akkadotnet/Akka.Persistence.MongoDB#serialization - there was a big shift between 1.3* and 1.4 but we added a feature toggle to make the system backwards compatible where needed.

**Reply by merthamurcu** [reply]
I check that in first days of testing. people adding new fields to payload objects, they now causing "No matching creator found", adding BsonDefaultValueAttribute solves some cases but can't give default to such as guid because attribute just doesn't suitable to that, can't even convert from string to guid I guess. so I think I manage to work it with IMemberMapConvention and set default values to each member, still early to say if its working. I can't be the first one who faces these issues?

**Reply by Aaronontheweb** [reply]
We worked with other long-time Akka.Persistence.MongoDb customers at the time these changes were made (i.e. users who've been using it for 5+ years) and haven't heard of any problems like this before. We definitely want to get to the bottom of it and help you though.

**Reply by merthamurcu** [reply]
I would love to get some help from your expertise thb, do you have a discord channel etc that I can go into detail without keeping here busy?

**Reply by Aaronontheweb** [reply]
You can use the Gitter chat room: https://gitter.im/akkadotnet/akka.net but the best way to get some help from us is probably https://petabridge.com/contact/

**Reply by merthamurcu** [reply]
sent you an email to your personal web mail, thanks for the interest
